### DISCUSSION TITLE
Job Scheduling in For Loop

### DISCUSSION THREAD

**Discussion by mertmr** [jobrunr/jobrunr#263] [Q&A]
Hi, I don't know if I should open an issue for this but starting from here seemed like a good idea. In Kotlin when I schedule jobs in a for loop, parameters stay the same. I guess it would be better to tell the story in code:

`import org.jobrunr.configuration.JobRunr
import org.jobrunr.scheduling.JobScheduler
import org.jobrunr.storage.InMemoryStorageProvider

fun main() {
    val jobScheduler: JobScheduler = JobRunr
        .configure()
        .useStorageProvider(InMemoryStorageProvider())
        .useBackgroundJobServer()
        .useDashboard()
        .useJmxExtensions()
        .initialize()
        .jobScheduler

    for (i in 1..10) {
        jobScheduler.enqueue { println(i) }
    }
}
`

As you can guess I would like the jobs to print 1 to 10. But it will only print 1 all the time. Is this the correct behavior? What should I do?

**Comment by mertmr** [comment]
I believe now that this is an issue. After much debugging, I discovered that with the 4. version, there is a caching mechanism for job detail retrieval. It is a nice mechanism for fast job detail retrieval but using cache key for only the class name(to be specific it is lambda.getClass()) confuses the mechanism. For my example above, the first enqueued jobs class puts into the cache with its class name and the second one after that job has the same class(literally it is the same thing with a different parameter) so the cache picks up the same jobDetail with the first one's parameter. 

I reverted to 3.2.1 which doesn't have this mechanism and it just works as intended. Am I doing something wrong with this cache mechanism? Any insight would be appreciated :)

**Reply by rdehuyss** [reply]
Hmmm, that can definitely be the reason. Sorry for that.

I'm a little bit short on time now but I think if you use real jobs (not sysout) you will not encounter this issue. Or do you?

**Reply by mertmr** [reply]
Actually, first I encountered this problem using real jobs. After that, I tried something simple like print to identify the issue. It is the same problem all around.

I am very grateful for this excellent project so if I can help you in any other way I am ready :)

**Comment by rdehuyss** [comment]
You can also turn the caching off by means of configuration. 

Really strange, this is now live for over a month and you are the first one to encounter it, that's why I didn't think of it.

**Reply by mertmr** [reply]
I couldn't see how to do that with a quick look at the documentation. Can you elaborate?

**Reply by mertmr** [reply]
I discovered something looking at the javadocs, is it the .useJobDetailsGenerator(JobDetailsAsmGenerator()) ?

**Reply by rdehuyss** [reply]
Yes, that's it. By leaving out the caching wrapper class, you should not encounter this.

**Reply by mertmr** [reply]
Thanks, this made things right :)
